### ISSUE TITLE
Feature Request: Add Globbing Pattern Support for Files in .slnx

### ISSUE THREAD

**Issue by kasperk81** [microsoft/vs-solutionpersistence#61]
**Feature Request: Add Globbing Pattern Support for Files in `.slnx`**

I'd like to suggest adding support for globbing patterns in `.slnx` files, similar to how it's done in `.csproj` using the `Microsoft.Build.Globbing` package. This would make managing large solutions simpler, especially when adding multiple files.

For example:

```xml
<File Path="test\**\*.cs" />
```

This would help dynamically include files without needing to list each one individually, streamlining file management.

For more context, please refer to the [Microsoft documentation on file globbing](https://learn.microsoft.com/en-us/dotnet/core/extensions/file-globbing).

**Comment by alrz** [comment]
Following sdk-style project format I think this should be the default for `csproj` files because the solution structure is usually in sync with file system, including any subfolders. 
```cs
<Solution Sdk="Microsoft.NET.Sdk">
  <!-- Includes every csproj by default -->
   <Project Remove="Project.csproj"/>
  <!-- Additional folders etc -->
</Solution>
```
Right now there must be an additional definition per each directory.

**Comment by jnm2** [comment]
This would solve a longstanding pain point. Most times when I add a new file to a solution folder, I have to create it separately and then add as an existing file as a separate step. It's very manual and it's easy to forget to keep it in sync. It would be incredibly nice if the solution folder would dynamically update.

**Comment by alrz** [comment]
Assuming globbing is in place to reflect the file system (perhaps by default), I think for the most part the sln file becomes trivial `<Solution/>`. I wonder if it could be only a tooling feature where the ide operates on a directory - opening csproj would open the git reposition root, for instance.

**Comment by pebezo** [comment]
This is the missing feature! The current solution behavior is to create a "virtual view" over how files and projects are actually laid out on disk. Especially as a solution grows, keeping track of the reality on disk becomes harder and more disjointed. Yes, we have the "file system" view, but that's a hack at best since we lose a number of benefits from the solution view. 

Having something like this would allow us to poke holes (portals?) into this virtual view. I find that the root folder is usually the most problematic. All the .gitignore, .yml, docker files etc. end up there because other 3rd party tool require them to be there, but VS (and other .NET IDEs) make us leave the IDE and/or fight against us being able to do this.

**Comment by robertmclaws** [comment]
Yes, please add this feature! Having to have a node per file in a complex path structure is... cumbersome.

**Comment by lukasvosyka** [comment]
> Supporting globbing, would require scanning the file system and evaluating potentially complex patterns to identify matching projects. Especially for larger repositories, this additional step can significantly slow down the loading process which is why globbing is not currently supported.

I think too that this would be a helpful feature in some scenarios such as mono repos. Having many projects and having the need to also maintain the .slnx file is cumbersome when you could just have set up rules. Please do not disallow something useful, just because you think it might result in slow loading times. Let the developers/team decide what the right thing is for them.

**Comment by tats-u** [comment]
This is especially helpful for solution items.

```xml
<Solution Sdk="...">
  <!-- Includes every csproj, directory props, targets by default -->
  <Project Remove="Project.csproj"/>
  <Folder Path="Solution Items"> <!-- Could be implicit as well? Why would you not want to include this? -->
     <!-- Includes all files in the root -->
     <File Path="*.md" />
     <File Path="LICENSE*" />
     <File Path=".editorconfig" />
      <!-- .gitignore & .gitattributes -->
     <File Path=".git*" />
  </Folder>
</Solution>
```

**Comment by kasperk81** [comment]
We proposed a new constructor to preserve the include/exclude order in the `Matcher` ctor, which has been included in the 10.0 package of `FileSystemGlobbing`:  
https://learn.microsoft.com/dotnet/api/microsoft.extensions.filesystemglobbing.matcher.-ctor?view=net-10.0-pp#microsoft-extensions-filesystemglobbing-matcher-ctor(system-stringcomparison-system-boolean)

Setting `preserveFilterOrder: true` makes it respect the user-defined order in the `.slnx` file: WYSIWYG.

Note: The 10.0 package remains compatible with `netstandard2.0`:  
https://www.nuget.org/packages/Microsoft.Extensions.FileSystemGlobbing#supportedframeworks-body-tab

**Comment by jasoncouture** [comment]
I added some push back on the performance argument on [PR #139](https://github.com/microsoft/vs-solutionpersistence/pull/139#issuecomment-4391727327), linking it here for visibility.

EDIT: The pushback includes benchmarks showing < 1s timings on a massive tree, the largest solution I've worked on in my entire 22 year career.

**Comment by thomaslevesque** [comment]
Has there been any progress on this? Is it even planned?

**Comment by jzabroski** [comment]
@baronfel We just need git fnmatch-style - we do not need globbing. This closely resembles what other systems like Rust's cargo toml system does and various node glob tools do. In general supporting globbing for whole solutions is an anti-pattern and even Google Bazel emits lint warnings if you do **\* style `glob()` directives.

One feature from Google Bazel worth carrying over is allow_empty parameter to glob(), which allows failing if the glob returns an empty result. This helps catch innocent typos, such as fat fingering a merge conflict fix and not realizing it.

**Comment by kasperk81** [comment]
Disagree. A lint warning of some unrelated tool is irrelevant. What's relevant for VS is SDK-style projects are using globbing in VS since 2017 with great success.

**Comment by jzabroski** [comment]
Would my solution here work for you? https://github.com/microsoft/vs-solutionpersistence/pull/139#issuecomment-4930329700

1. Use Microsoft.Build.NoTargets SDK
2. Glob inside that csproj

**Comment by jods4** [comment]
@jzabroski beyond solutions files, I would really like to see globing work for projects:
```xml
<Project Path="Modules/**/*.csproj" />
```

**Comment by jzabroski** [comment]
Do you ever have projects that are multiple levels deep?

I typically see

```
/Source/<ProjectName>/<ProjectName>.csproj
```

or

```
/Source/Modules/<ProjectName>/<ProjectName>.csproj
```

It seems like something like:

```
<ProjectRoot Path="/Source/Modules/" Depth="1" /> <!-- Depth Would be optional and default to 1 -->
```

Would be preferred. Because it's not going to scan the entire tree.

**Comment by CZEMacLeod** [comment]
@jzabroski 

> Do you ever have projects that are multiple levels deep?

Yes - often multiple different levels in several branches. I've got various solutions with >200 projects like that.

> I typically see
> 
> ```
> /Source/<ProjectName>/<ProjectName>.csproj
> ```
> 
> or
> 
> ```
> /Source/Modules/<ProjectName>/<ProjectName>.csproj
> ```

`/src/VerticalSlice/Each/Part/Of/A/NameSpace/ProjectType/Each.Part.Of.A.NameSpace.ProjectType.csproj`

Where `ProjectType` would be something like ObjectModel, BusinessLogic, Persistance, API, Web, etc.

And since there may be projects at different levels of a namespace or plugins/providers etc. projects often have different levels of nesting.

What is true is that if you find a project, you don't need to scan sub-folders of that project.

Also `/test/....` projects that match up.

> It seems like something like:
> 
> ```
> <ProjectRoot Path="/Source/Modules/" Depth="1" /> 
> ```
> 
> Would be preferred. Because it's not going to scan the entire tree.

```xml
<Folder Path="src">
  <Project Path="/src/**/*.csproj" />
  <Project Path="/src/**/*.vbproj" />
</Folder>
<Folder Path="test">
  <Project Path="/test/**/*.csproj" />
  <Project Path="/test/**/*.vbproj" />
</Folder>
```

Check something like [C3D.Extensions.Aspire](https://github.com/CZEMacLeod/C3D.Extensions.Aspire/blob/main/C3D.Extensions.Aspire.slnx)
for a very minimal example - a simple library repo. Full stack projects are often much more involved.

**Comment by kasperk81** [comment]
NoTargets approach and many similar ones (have a script OpenSolution.py generating slnx on the fly) are workarounds or hacks since there is no MSBuild evaluation involved in native VS solution loading.

**Comment by kieronlanning** [comment]
As a suggestion for globbing support over large projects, could a build warning be raised if the globbing takes too long? Like a function of the average load/ scan time vs. no of project/ matches found. 

Then everyone's happy - you can use until your hearts content, and ignore the warnings or remove globbing.

Choice is on the user. 

Then, over time performance is improved, the function's time series is extended until eventually it's removed.

**Comment by kasperk81** [comment]
> As a suggestion for globbing support over large projects, could a build warning be raised if the globbing takes too long?

Visual Studio already handles this gracefully via the solution loading bar. https://github.com/dotnet/roslyn runs MSBuild globbing over 20k individual source files at Roslyn.slnx load time. Locating a mere 400 project files via glob would be 50x smaller in scope.

**Comment by jods4** [comment]
@jzabroski 
> Do you ever have projects that are multiple levels deep?

Are you re-inventing an equivalent to the glob `/Source/Modules/*/*.csproj` ?

**Comment by jzabroski** [comment]
Yeah, pretty much 🤣 . I played around more with slnx today and realized that I have use cases for globbing that fnmatch doesn't really address. Mostly due to github copilot artifacts like skills, prompts, etc. the .github directory is exactly as @Czemacleod mentions

**Comment by MichelZ** [comment]
> Yeah, pretty much 🤣 . I played around more with slnx today and realized that I have use cases for globbing that fnmatch doesn't really address. Mostly due to github copilot artifacts like skills, prompts, etc. the .github directory is exactly as [@CZEMacLeod](https://github.com/CZEMacLeod) mentions

If you're looking for .github in your solution in visual studio, try [this](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.GitHubNode) extensions from Mads Kristensen

**Comment by bartlomiej-dawidow** [comment]
This would satisfy most of our needs:
```
<Folder Path="src">
  <Project Path="/src/*/*.csproj" />
</Folder>
<Folder Path="test">
  <Project Path="/test/*/*.csproj" />
</Folder>
```

Currently, whenever someone on the team adds, removes, or moves a project, they also need to touch the `.slnx`. This adds unnecessary work and boilerplate, and occasionally causes merge conflicts.

It also creates a risk that the `.slnx` will become out of sync with the actual folder structure - which we never want.

**Comment by jods4** [comment]
With 135 👍 at the time of this comment, this issue is easily the most upvoted issue in this repo: it has almost 4x as many reactions as the 2nd issue!

Looks like the community is strongly interested in this... please find a way 🙏
I believe the example from Bartlomiej above is what most people want.

**Comment by jzabroski** [comment]
It seems like the current blocker to supporting this might be https://github.com/dotnet/msbuild/issues/14420, assuming vs-solutionpersistence would use the same Glob Syntax/Features as [IMSBuildGlob](https://github.com/dotnet/msbuild/blob/9a0cef6f75bf13ffbbde956b8f7d7ad7d6e0d996/src/Build/Globbing/IMSBuildGlob.cs#L24), which makes a ton of sense to me. But it sounds exciting - @jeremykuhne 's Touki library would address a lot of the performance concerns with globbing, and could even replace vscode's file watcher component in the future as TypeScript-based one does not have the same ability to reason about allocation-free data structures (no `Span` data type).
